pymbolic/primitives.py

```diff
@@ -681,6 +720,116 @@ def __iter__(self):
         raise TypeError("expression types are not iterable")
 
 
+# {{{ dataclasses support
+
+def augment_expression_dataclass(cls) -> Type:
+    """A class decorator for :func:`dataclasses.dataclass`-derived
+    :class:`Expression` nodes. It adds cached hashing, equality comparisons
+    with ``self is other`` shortcuts as well as some methods/attributes
+    for backward compatibility (e.g. ``__getinitargs__``, ``init_arg_names``)
+    """
+    attr_tuple = ", ".join(f"self.{fld.name}" for fld in fields(cls))
+    if attr_tuple:
+        attr_tuple = f"({attr_tuple},)"
+    else:
+        attr_tuple = "()"
+
+    fld_name_tuple = ", ".join(f"'{fld.name}'" for fld in fields(cls))
+    if fld_name_tuple:
+        fld_name_tuple = f"({fld_name_tuple},)"
+    else:
+        fld_name_tuple = "()"
+
+    comparison = " and ".join(
+            f"self.{fld.name} == other.{fld.name}"
+            for fld in fields(cls))
+
+    if not comparison:
+        comparison = "True"
+
+    from pytools.codegen import remove_common_indentation
+    augment_code = remove_common_indentation(
+        f"""
+        from warnings import warn
+
+
+        def {cls.__name__}_eq(self, other):
+            if self is other:
+                return True
+            if self.__class__ is not other.__class__:
+                return False
+            if hash(self) != hash(other):
+                return False
+            if self.__class__ is not cls and self.init_arg_names != {fld_name_tuple}:
+                warn(f"{{self.__class__}} is derived from {cls}, which is now "
```

Comment:
```suggestion
                warn(f"{{self.__class__}} is derived from {cls.__name__}, which is now "
```
? Otherwise that should print `<class 'Expression'>` or something like that (?)

Reply:
I actually think that's OK... while awkward-looking, it is fully qualified at least.